packages/pool/src/sim/index_in.rs

```diff
@@ -0,0 +1,181 @@
+                        routes: None,
+                    })
+                } else {
+                    self.estimate_out_given_in(token_in, output_asset, &mut pools_map)
+                }
+            })
+            .collect::<Result<Vec<_>, PoolError>>()?;
+
+        let total_received = routes_with_amount
+            .iter()
+            .try_fold(Uint128::zero(), |acc, v| {
+                Ok::<_, PoolError>(acc.checked_add(v.sim_amount_out)?)
+            })?;
+
+        let ret = SimIndexInResp {
+            total_received,
+            sim_routes: SimAmountOutRoutes(routes_with_amount),
+        };
+
+        Ok(ret)
+    }
+
+    fn search_efficient_amount_for_output_f(
+        &self,
+        desired_output: Coin,
+        token_test_out: Uint128,
+        token_test_in: Uint128,
+    ) -> Result<SearchAmountForOutputResp, PoolError> {
+        let est_min_token_in =
+            token_test_in * Decimal::checked_from_ratio(desired_output.amount, token_test_out)?;
+
+        let res = self.estimate_token_given_index_in(est_min_token_in, &desired_output.denom)?;
+
+        let ret = SearchAmountForOutputResp {
+            est_min_token_in,
+            max_token_out: res.total_received,
+            sim_routes: res.sim_routes,
+        };
+
+        Ok(ret)
+    }
+
+    pub fn search_efficient_amount_for_output(
+        &self,
+        desired_output: Coin,
+        init_input_amount: Option<Uint128>,
+    ) -> Result<SearchAmountForOutputResp, PoolError> {
+        let mut acc_res = {
+            let token_in = init_input_amount.unwrap_or(Uint128::new(1000000));
+            let est_res = self.estimate_token_given_index_in(token_in, &desired_output.denom)?;
+
+            SearchAmountForOutputResp {
+                est_min_token_in: token_in,
+                max_token_out: est_res.total_received,
+                sim_routes: est_res.sim_routes,
+            }
+        };
+
+        let mut loop_count = 0;
+        loop {
+            if loop_count >= MAX_LOOP {
+                return Err(PoolError::MaxLoopExceeded);
+            }
+
+            acc_res = self.search_efficient_amount_for_output_f(
+                desired_output.clone(),
+                acc_res.max_token_out,
+                acc_res.est_min_token_in,
+            )?;
+
+            let err = Decimal::checked_from_ratio(desired_output.amount, acc_res.max_token_out)?;
+            // self.deps.api.debug(&format!(
+            //     "loop {}. input_amount={}, token_in={}, token_out={}, err={}",
+            //     loop_count,
+            //     desired_output.amount,
+            //     acc_res.est_min_token_in,
+            //     acc_res.max_token_out,
+            //     err
+            // ));
+
+            if Decimal::one() <= err
+                && desired_output.amount - acc_res.max_token_out < Uint128::from(MAX_ERROR)
+            {
+                return Ok(acc_res);
+            }
+            loop_count += 1;
+        }
+    }
```

Comment:
This logic can be replaced to `for loop`. In that cases, you should choice the `for loop` for prevent unexpected infinite loop

Reply:
ah okay got it